### PR TITLE
Filtering by arbitrary expressions

### PR DESCRIPTION
**Pull request by kornholi**
I'm thinking about moving everything from `metadata_filtering` into `dl.utilities`, that way you can do `from dl.utilities import fields as f`. Thoughts? Maybe only put the actual `Fields` object there?

Will add docs once we decide on the final syntax.

### CODE REVIEW

File: descarteslabs/services/metadata_filtering.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+    def __getattr__(self, attr):
+        if attr in self.fields:
+            return Variable(attr)
+
+        raise AttributeError(
+            "'Fields' object has no attribute '{}'".format(attr))
+
+
+fields = Fields("absolute_orbit", "acquired", "archived", "area", "azimuth_angle",
```
Comment (justinwp) [inline descarteslabs/services/metadata_filtering.py]: need a product attribute

File: descarteslabs/services/metadata_filtering.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+class AndExpression(object):
+    def __init__(self, parts):
+        self.parts = parts
+
+    def __and__(self, other):
+        if isinstance(other, AndExpression):
+            self.parts.extend(other.parts)
+            return self
+        if any(isinstance(other, ty) for ty in [OrExpression, Expression]):
```
Comment (justinwp) [inline descarteslabs/services/metadata_filtering.py]: `if isinstance(other, (OrExpression, Expression,)):`

File: descarteslabs/services/metadata_filtering.py (hunk 1 above)
Comment (hrfuller) [inline descarteslabs/services/metadata_filtering.py]: products

File: descarteslabs/services/metadata.py [hunk 3]
```diff
@@ -175,19 +175,21 @@ def summary(self, products=None, const_id=None, sat_id=None, date='acquired', pa
         if fill_fraction is not None:
             kwargs['fill_fraction'] = fill_fraction
 
+        if filter is not None:
+            if not isinstance(filter, list):
```
Comment (hrfuller) [inline descarteslabs/services/metadata.py]: Probably shouldn't clobber builtin filter method?

**Comment by kornholi** [comment]
Anything left to do here? Right now the syntax is:

```py
import descarteslabs as dl
from descarteslabs.utilities import properties as p

dl.metadata.search(q=[(p.cloud_fraction < 0.5) | (p.cloud_fraction_1 < 0.5)])
```

I'm thinking we could figure out docs in an another PR

**Comment by kornholi** [comment]
Added some docs

**Comment by kornholi** [comment]
@samskillman @justinwp

**Comment by kornholi** [comment]
params is now gone

**Pull request merged**